### ISSUE TITLE
DCP check fails on a positive definite quadratic form (ARPACK convergence failure)

### ISSUE THREAD

**Issue by jomoussafir** [cvxpy/cvxpy#1995]
**Describe the bug**

I did solve two equivalent optimisation problems using cvxpy with default solver, cvxopt or mosek and I observed each time 
the same pattern. The two problems are equivalent in the following sense, the second one corresponds to the first one with a simple change of variable. The second problem is quickly solved, but the first one either fails or takes a long time to run.
When both succeed, solutions match.

**To Reproduce**
The following python code reproduces what I got :

```
import cvxpy as cp
import numpy as np
np.set_printoptions(precision=2)
np.set_printoptions(linewidth=200)

import matplotlib.pyplot as plt

print(f"cvxpy: {cp.__version__}")
print(f"numpy: {np.__version__}")

nb_ineq = 10
nb_var = 200

np.random.seed(1)

P = np.random.randn(nb_var, nb_var)
P = P.T @ P

U = np.linalg.cholesky(P).T
U_inv = np.linalg.inv(U)

a = np.random.randn(nb_var)
G = np.random.uniform(low=-10, high=10, size=nb_ineq*nb_var).reshape((nb_ineq, nb_var))
h = np.random.uniform(low=1, high=10, size=nb_ineq)

## initial problem
x = cp.Variable(nb_var)
prob = cp.Problem(cp.Maximize(a.T@x - (1/2)*cp.quad_form(x, P)), [G @ x <= h])
prob.solve()
x_sol = x.value

## transformed problem
y = cp.Variable(nb_var)
prob_tr = cp.Problem(cp.Maximize((a.T @ U_inv)@y - (1/2)*cp.quad_form(y, np.eye(nb_var))), 
                  [(G @ U_inv) @ y <= h])
prob_tr.solve()
x_tr_sol = U_inv @ y.value
```
The run time for the first problem can be several orders of magnitude larger than the run time for the second one.
Most of the time with large values for nb_var, the first problem will fail with error ending like this
`ArpackNoConvergence: ARPACK error -1: No convergence (2001 iterations, 0/1 eigenvectors converged)

**Expected behavior**
I would expect both problem to converge and use approximately the same time. 
The problem is feasible here since $h\geq0$.

**Output**
If applicable, include program output. If reporting a program crash, please include the entire stack trace.

**Version**
 - OS: Darwin Kernel Version 20.3.0: Thu Jan 21 00:07:06 PST 2021; root:xnu-7195.81.3~1/RELEASE_X86_64 x86_64
 - CVXPY Version: 1.2.3

I have run the same code on a linux box with same version of cvxpy, and notice the same problem.

**Comment by aszekMosek** [comment]
You should not expect the same performance since in the first method cvxpy is required to compute a (Cholesky or similar) factorization of a large, dense ``P`` for use in ``quad_form``, whereas in the second method you essentially provide this factorization in advance by only using  ``U``.

As to why the computation of the factorization fails I will leave to someone more knowledgeable about ``scipy``. 

Apart from that, your second model will be much faster because ``nb_ineq`` is very small compared to ``nb_var`` and therefore it has much fewer nonzeros (of order ``nb_ineq*nb_var`` rather than ``nb_var*nb_var``)

**Comment by rileyjmurray** [comment]
@aszekMosek raises some good points. Although, CVXPY supports some solvers that operate on quadratic objectives natively, and so CVXPY isn't factoring ``P`` directly in this case. What's happening is that CVXPY is attempting to verify that ``P`` is PSD as part of the DCP check. In an attempt to _avoid_ factorization, we use an iterative method to compute the smallest eigenvalue in checking that a matrix is PSD. That iterative method can fail for numerical reasons.

You can fix this issue by using the following objective function
```
objective = cp.Maximize(a.T@x - (1/2)*cp.quad_form(x, P_psd))
```
where you set ``P_psd`` as follows in CVXPY 1.2.x
```
P_psd = cp.atoms.affine.wraps.psd_wrap(P)
```
or as follows in CVXPY 1.3.x:
```
P_psd = cp.psd_wrap(P)
```

I think there is something we can do to help here. We should update the ``is_psd_within_tol`` function in  ``cvxpy/utilities/linalg.py`` (as of 1.3.x, old file is ``cvxpy/utilities/eigvals.py``) so that if an ARPACK convergence failure occurs then we recommend someone consider the ``psd_wrap`` atom.  @phschiele @PTNobel what do you two think?

**Comment by phschiele** [comment]
@rileyjmurray I agree that we should add a descriptive error message here. 
We can even use the `assume_PSD` argument of `quad_form`:
https://github.com/cvxpy/cvxpy/blob/8fae482f0afa3d6f8395039d6a46a25839a4638c/cvxpy/atoms/quad_form.py#L225
I can pick this up.

**Comment by PTNobel** [comment]
I concur with Philipp, am happy to pick this up if he doesn't have time.

We need to be careful how we write the error message; if it is only for when ARPACK convergence fails, does that suggest we're numerically close enough to PSD that we don't really care? And does that it mean it should have a low false-suggestions-to-use-psd_wrap rate?

**Comment by jomoussafir** [comment]
Thank you so much for your quick replies.
`P_psd = cp.atoms.affine.wraps.psd_wrap(P)
does fix the error.

On the computation time, I'm having the impression that the second problem - with cholesky decomposition computation time included - is always faster than the first one. I randomized a bit the tests using for instance

`nb_ineq in range(100, 1000)
`nb_var in range(100, 1000)

The speedup ranges approximately from factor 1 to 10 or more.

**Comment by aszekMosek** [comment]
Some unsorted comments. I used Mosek.

* Problem 2 has (order of) ``nb_ineq*nb_var`` nonzeros and Problem 1 has ``nb_var*nb_var`` and therefore the solver is much faster on problem 2 when ``nb_ineq`` is small compared to ``nb_var``. However timing everything, including Cholesky and matrix inversion, makes the gap much smaller.
* However if ``G`` was very sparse, as it typically happens, then ``G*U_inv`` will destroy sparsity, so the above is not simply a function of dimensions and will shift the balance. Benchmarking on a fully dense problem may not reflect the reality of a real problem you want to solve.
* ``cvxpy`` seems to compute the factorization ``P=U.T@U`` using eigenvalue decomposition (if I correctly see, I leave it to the experts here), in particular the problem finally passed to Mosek contains ``Ux`` with a full square matrix ``U``, whereas using the Cholesky decomposition makes ``U`` triangular, so smaller (half as many nonzeros), and also may be easier for the solver to exploit in presolve. For example much more of the Mosek time in Problem 1 is spent in presolve because of this difference.
* details may vary depending on the solver, its presolve, and the fact that ``cvxpy`` actually passes the solver the dual. 

My favorite variant would be

```
	U = np.linalg.cholesky(P).T
	
	x = cp.Variable(nb_var)
	prob = cp.Problem(cp.Maximize(a.T@x - (1/2)*cp.sum_squares(U @ x)), [G @ x <= h])
	prob.solve(verbose=True)
	x_sol = x.value
```

It uses triangular ``U``, no inverse, no need to rely on ``cvxpy`` to factorize, not affecting ``G``, good speed.

Sorry for the offtopic. A modeling discussion and me ranting about the hidden dangers of ``quad_form`` should maybe take place elsewhere ;).

**Comment by rileyjmurray** [comment]
Not at all off-topic, @aszekMosek! Your comments here are always welcome.

Our rationale for using eigendecomposition is that it works with singular matrices. Cholesky is certainly faster than eigendecomposition, but I expected the difference would be marginal compared to what happens inside a numerical solver. I didn't realize that dense versus triangular ``U`` in ``sum_squares(U @ x)`` could make a difference in presolve.

The ideal canonicalization codepath for conic solvers would probably start with Cholesky and fall back on eigendecomposition if needed. The decomposition could be used as the method for checking if a matrix is PSD (rather than an iterative method in ARPACK). If the matrix ``P`` is sparse then ideally we'd do sparse Cholesky. Right now SciPy doesn't expose that functionality so we would need to access that from Eigen.

**Comment by rileyjmurray** [comment]
Resolved by #2125.
